### DISCUSSION TITLE
Cross-Platform iOS not supporting Save file picker

### DISCUSSION THREAD

**Discussion by Rushikesh3009** [AvaloniaUI/Avalonia#13641] [Q&A]
Hi everyone,

I am trying to save file in iOS Files using following code

```
  TopLevel topLevel = TopLevel.GetTopLevel(this);
  var file = await topLevel.StorageProvider.SaveFilePickerAsync(new FilePickerSaveOptions
  {
      Title = "Save Text File",      
  });
```

This is cross-platform code as it is mansion in avalonia website: 
https://docs.avaloniaui.net/docs/next/basics/user-interface/file-dialogs

This is working well on my Windows and Android platforms but it not working on iOS after digging the code I get this following message:

![img](https://github.com/AvaloniaUI/Avalonia/assets/149040605/57e35006-03eb-48d8-b4c5-ae6b1d05c381)

Is there any common code in avalonia which is useful to save file in Files app for all platform.

**Comment by workgroupengineering** [comment]
report InnerExceptions

**Reply by thevortexcloud** [reply]
It will always throw an exception based on what I am seeing when on iOS. It has not been implemented for some reason.

https://github.com/AvaloniaUI/Avalonia/blob/ea64505600bdedce055b35f391c85ec1056e359d/src/iOS/Avalonia.iOS/Storage/IOSStorageProvider.cs#L141-L145

**Comment by thevortexcloud** [comment]
I wonder why this is not supported. The open file picker is apparently implemented. From what I can see iOS as a platform does allow for using a save file picker (although I may be missing things). Avalonia just for some reason has not implemented it? Xamarin looks like it can do it. Which means you can probably go via Xamarin if you really need to as Avalonia uses it for mobile support anyway.

https://learn.microsoft.com/en-us/xamarin/ios/platform/document-picker

**Reply by maxkatz6** [reply]
I don't see any mentioning of *save* file picker API there.

**Comment by maxkatz6** [comment]
There is no *save* file picker on the iOS.
You can pick a folder, and create file inside, and then write to it. 

Something like this:
```cs
IStorageFile file;
if (storage.CanSave)
{
     file = await storage.SaveFilePickerAsync();
}
else
{
    var folder = await storage.OpenFolderPickerAsync();
    file = await folder.CreateFileAsync("file.txt");
}
await using var writeStream = file.OpenWriteAsync();
// write to writeStream.
```

**Reply by maxkatz6** [reply]
Also: https://docs.avaloniaui.net/docs/next/concepts/services/storage-provider#platform-compatibility
